testsuite/tests/weak-ephe-final/finaliser2.ml

```diff
@@ -1,39 +1,47 @@
 (* TEST *)
 
+(* If minor heap values are finalised on a domain, then the callbacks are run
+   at the end of a minor collection triggered by that domain. *)
 let[@inline never][@local never] test1 () =
   let r' = ref 0 in
   let rec foo () =
     let r = ref 0 in
-    Gc.finalise_last (fun _ -> r' := 1; print_endline "test1") r
+    Gc.finalise_last (fun _ -> r' := 1; Printf.printf "test1\n%!") r
   in
   foo();
   Gc.minor();
   assert (!r' = 1)
 
+(* Check that the finalisation functions will be called in the reverse order of
+   the corresponding calls to finalise. *)
 let[@inline never][@local never] test2 () =
   let r = ref 0 in
-  Gc.finalise (fun r -> assert (!r = 1); print_endline "test2: 1") r;
-  Gc.finalise (fun r -> assert (!r = 0); print_endline "test2: 2"; r := 1) r;
+  Gc.finalise (fun r ->
+    assert (!r = 1); Printf.printf "test2: installed 1st\n%!") r;
+  Gc.finalise (fun r ->
+    assert (!r = 0); r := 1; Printf.printf "test2: installed 2nd\n%!") r;
   Gc.full_major()
 
+(* The finalisation functions attached with Gc.finalise are always called
+   before the finalisation functions attached with Gc.finalise_last. *)
 let[@inline never][@local never] test3 () =
   Gc.full_major ();
-  let rec foo () =
-    let r = ref 0 in
-    Gc.finalise (fun r -> print_endline "test3: parent.1") r;
-  in
-  foo ();
```

Comment:
This part of the test is definitely uninteresting, right?

Reply:
Yes, indeed. The deleted bits are uninteresting for this particular test. We know that finalisers installed on the same domain work as expected thanks to “test1” function (and also “weak-ephe-final/finalisers.ml” test).